### ISSUE TITLE
[PT-03] DPO Direct Preference Optimization (Rafailov 2023)

### ISSUE THREAD

**Issue by jsboige** [jsboige/CoursIA#1756]
## Parent
Epic #1742 (RL/Post-training SOTA — Deepseek-Qwen style)
Epic #1454 (Training & Post-Training)

## Scope

Notebook `MyIA.AI.Notebooks/PostTraining/PT_03_dpo_direct_preference.ipynb` : implementer **Direct Preference Optimization** (Rafailov et al., NeurIPS 2023) sur Qwen2.5-0.5B post-SFT (deja livre via PT-02 PR #1755 merged).

DPO contourne le reward modeling explicite de RLHF en optimisant directement sur des paires de preference (chosen vs rejected). Formule cle :

```
L_DPO = -E[log sigmoid(beta * (log pi_theta(y_w|x) - log pi_ref(y_w|x)) - beta * (log pi_theta(y_l|x) - log pi_ref(y_l|x)))]
```

ou `y_w` = chosen (preferred), `y_l` = rejected, `pi_ref` = policy de reference (modele SFT gele).

## Implementation

- Library : `trl.DPOTrainer` (HuggingFace TRL)
- Base model : checkpoint SFT de PT-02 (ou load equivalent depuis HF si SFT non-realise)
- Dataset : `HuggingFaceH4/ultrafeedback_binarized` (train_prefs split, fields chosen/rejected)
- LoRA : meme config que PT-02 (r=8, alpha=16, attention+MLP)
- Hyperparameters cles : beta=0.1 (KL regularization strength), learning rate 5e-7 (typique DPO, plus bas que SFT)
- Reference model : pi_ref clone du modele SFT en mode `requires_grad=False`

## Cellules attendues

1. Math du loss DPO (LaTeX, intuition Bradley-Terry → preference logit)
2. Pourquoi DPO contourne reward modeling explicite (avantage vs RLHF classique)
3. Setup datasets (ultrafeedback chosen/rejected pairs)
4. Tokenization + preprocessing DPO format
5. Charger checkpoint SFT comme policy + clone comme reference
6. `DPOConfig` + `DPOTrainer` construction
7. Conditional training avec `LOAD_MODEL_AND_TRAIN` flag pattern (cf PT-02)
8. Eval preference accuracy : freq(pi_theta prefere chosen) sur subset eval
9. 4 piege classiques DPO : KL collapse, reference drift, overfit beta, format token chat
10. Transition vers PT-04 (GRPO)

## Acceptance criteria

- C.1 : 0 NotImplementedError/assert False
- C.2 : execute end-to-end avec outputs reels committes (avec ou sans entrainement reel, flag explicite)
- Multi-seed >= 4 si claim "DPO > SFT" sur eval (sinon flag INCONCLUSIVE honnete)
- Reference Rafailov NeurIPS 2023 + lien arXiv 2305.18290

## Contraintes RTX 3070 8 Go

- Quantization 4-bit obligatoire (bitsandbytes NF4 + double quant)
- batch size 1-2 + ga 16 (DPO double la memoire vs SFT car pi_ref load)
- max_seq_length <= 1024
- training time cible : < 90 min

## References

- Rafailov et al., "Direct Preference Optimization: Your Language Model is Secretly a Reward Model" (NeurIPS 2023), arXiv:2305.18290
- Tunstall et al., "Zephyr: Direct Distillation of LM Alignment" (2023) — exemple end-to-end SFT+DPO
- HuggingFace TRL DPOTrainer docs

## Machine

po-2024 (RTX 3070 8 Go) pioneering. Side-track Epic #1742.

## Priorite

Sub-issue #1742-C apres PT-02 (#1742-B livre PR #1755). DPO = brique avant GRPO (PT-04 livrable cle).

**Comment by jsboige** [comment]
Delivered via PR #1766 (#1756), #1768 (#1757), #1771 (#1758), #1772 (#1759). All merged to main. Epic #1742 COMPLETE.
